### ISSUE TITLE
Psalm does not detect when a function returning float returns an int

### ISSUE THREAD

**Issue by ShadauxCat** [vimeo/psalm#5538]
https://psalm.dev/r/9d8650a324

When a function is documented as a float and returns a variable that's determined to be an int, psalm should throw an error. As seen in the example, the documentation that the function returns a float can lead to runtime errors when the returned int value is === checked against a float literal, because `0 !== 0.0` in php.

**Comment by orklah** [comment]
This is slightly by design and results from PHP weirdness
For example, being strict between float and int would force casting to int every result of ceil/floor/round: https://3v4l.org/05p0W

I think there may be a way to differentiaite between float emitted from functions like those and phpdoc (with an internal property from_calculation) but it's kinda above my current level and understanding of psalm

**Comment by ShadauxCat** [comment]
I kind of already have to do those casts. For example: `if(floor(0.5) === 0)` will fail because `floor()` returns a float. Also, I would argue it's a little inconsistent to be extremely strict on types and throw an error on `2 * 2.0` due to adding int to float (or in this case, multiplying, but the error still says adding) but be loose with types in function return values.

Seems the current psalm rules treat passing an int as a float as acceptable but passing a float as an int as an error... but in reality, both are runtime errors in the right contexts.

Of course, the error goes away if I use PHP7 type annotations:

```php
    /** @return float */
    static function bar(): float
    {
        return static::$foo;
    }
```

Because then php will cast it to a float for me. But given that psalm supports much more powerful return type annotations than PHP does and I want to keep my codebase consistent (and also presume that those annotations incur a run-time cost that is unnecessary thanks to psalm), I haven't been using the PHP annotations.

**Comment by AndrolGenhald** [comment]
I think this is by design, since `strict_types=1` still allows `int` to be coerced to `float` without issue.
https://3v4l.org/S5guj
I think the solution here is to use PHP return types. I haven't seen any indication that there is any noticeable performance cost.
It might be possible to add a config option for this, but it's currently a feature, not a bug.

**Comment by ShadauxCat** [comment]
My issue with using PHP return types is less about runtime cost, and more about consistency. Since psalm return types are more powerful than php return types, I can't use php return types everywhere. If I can't use them everywhere, I'd rather not use them anywhere.

I'd also point out that psalm (at least in the configuration I'm using - not the one on psalm.dev though) actually DOES flag errors with return values from floor()/round()/etc. If I try to do something like `floor(1.5) * 5` I get an error about multiplying floats by ints.

I can certainly see the argument that not everyone would want to deal with that kind of strictness and having to cast things everywhere, so I certainly understand not making it default behavior to behave that way. But I run psalm at the maximum error level at all times... in that context, I would like to see an error flagged when returning an int from a function that returns float. So yes, a config option is what I'm asking for. :)

**Comment by AndrolGenhald** [comment]
> Since psalm return types are more powerful than php return types, I can't use php return types everywhere.

It should always be possible to use both, as long as the docblock return type from `@return` or `@psalm-return` is more specific than the php return type, psalm should just ignore the php return type. If there's a case where that's not working, it's a bug.

Of course, a config for this would be good as well.
